Spokestack/SpeechContext.swift

```diff
@@ -18,8 +19,87 @@ import Foundation
+    private var listeners: [SpeechEventListener] = []
+    /// Current error in the pipeline
+    internal var error: Error?
+    /// Current trace in the pipeline
+    internal var trace: String?
+    
+    /// Initializes a speech context instance using the specified speech pipeline configuration.
+    /// - Parameter config: The speech pipeline configuration used by the speech context instance.
+    @objc public init(_ config: SpeechConfiguration) {
+        self.configuration = config
+    }
+    
+    /// Adds the specified listener instance to the ordered set of listeners. The specified listener will recieve speech pipeline events.
+    ///
+    /// - Parameter listener: The listener to add.
+    @objc public func setListener(_ listener: SpeechEventListener) {
+        if self.listeners.contains(where: { l in
+            return listener === l ? true : false
+        }) { } else {
+            self.listeners.append(listener)
+        }
+    }
+    
+    /// Removes the specified listener by reference. The specified listener will no longer recieve speech pipeline events.
+    /// - Parameter listener: The listener to remove.
+    @objc public func removeListener(_ listener: SpeechEventListener) {
+        for (i, l) in self.listeners.enumerated() {
+            _ = listener === l ? self.listeners.remove(at: i) : nil
+        }
+    }
+    
+    /// Removes all listeners.
+    @objc public func removeListeners() {
+        self.listeners = []
+    }
+
+    @objc public func setStage(_ stage: SpeechProcessor) {
+        
+    }
+
+    @objc public func removeStage(_ stage: SpeechProcessor) {
+        
+    }
+
+    @objc internal func notifyListener(_ about: SpeechEvents) {
+        self.listeners.forEach { listener in
+            self.configuration.delegateDispatchQueue.async {
+                switch about {
+                case .initialize:
+                    listener.didInit()
+                case .start:
+                    listener.didStart()
+                case .stop:
+                    listener.didStop()
+                case .activate:
+                    listener.didActivate()
+                case .deactivate:
+                    listener.didDeactivate()
+                case .recognize:
+                    listener.didRecognize(self)
+                case .error:
+                    guard let e = self.error else {
+                        listener.failure(speechError: SpeechPipelineError.errorNotSet("A pipeline component attempted to send an error to SpeechContext's listeners without first setting the SpeechContext.error property."))
+                        return
+                    }
+                    listener.failure(speechError: e)
+                    self.error = .none
+                case .trace:
+                    guard let t = self.trace else {
+                        // swallow this error case because traces aren't important to the client.
```

Comment:
They might be (important). Best case here is that you're making app logging inconsistent based on what order the log listeners are attached—only the first listener will _ever_ get traces. I'd consider not clearing out the trace message immediately after sending. If you want it gone before the next event, wipe it after the loop has finished.

A bigger question here is whether it might make sense to adopt a more generic API for the listeners (like Android's single `onEvent` method) instead of having a separate method for each event. If this is the more idiomatic approach in Swift or iOS, so be it, but it does tend to make expansion a bit verbose.

Reply:
dispatch for individual events is idiomatic in iOS (back to the message passing days), so I'll accept the verboseness in this case.